File: src/main/java/de/spries/fleetcommander/model/ai/behavior/SearchEnemyHomePlanetFleetStrategy.java
package de.spries.fleetcommander.model.ai.behavior;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import de.spries.fleetcommander.model.core.universe.HasCoordinates;
import de.spries.fleetcommander.model.facade.PlayerSpecificPlanet;
import de.spries.fleetcommander.model.facade.PlayerSpecificUniverse;

public class SearchEnemyHomePlanetFleetStrategy implements FleetStrategy {

	@Override
	public void sendShips(PlayerSpecificUniverse universe) {
		PlayerSpecificPlanet homePlanet = universe.getHomePlanet();
		List<PlayerSpecificPlanet> allPlanets = universe.getPlanets();
		List<PlayerSpecificPlanet> myPlanets = PlayerSpecificPlanet.filterMyPlanets(allPlanets);
		Optional<PlayerSpecificPlanet> enemyHomePlanet = allPlanets.stream()
				.filter(p -> p.isKnownAsEnemyPlanet() && p.isHomePlanet()).findFirst();
		List<PlayerSpecificPlanet> unknownPlanets = allPlanets.stream()
				.filter(p -> !p.isInhabitedByMe() && !p.isKnownAsEnemyPlanet() && p.getIncomingShipCount() == 0)
				.sorted(HasCoordinates.getComparatorForClosestFirst(homePlanet))
				.collect(Collectors.toList());

		if(enemyHomePlanet.isPresent()) {
			// Full attack on enemy home planet
			for (PlayerSpecificPlanet myPlanet : myPlanets) {
				universe.sendShips(myPlanet.getShipCount(), myPlanet.getId(), enemyHomePlanet.get().getId());
			}
		}
		else {
			// Expand to next closest uninhabited planets
			for (PlayerSpecificPlanet unknownPlanet : unknownPlanets) {
				// send 1 ship from closest of my own planets which has ships > 0
				Optional<PlayerSpecificPlanet> closestOwnPlanetWhichHasShips = myPlanets.stream()
						.sorted(HasCoordinates.getComparatorForClosestFirst(unknownPlanet))
						.filter(p -> p.getShipCount() > 0)
						.findFirst();
				if(!closestOwnPlanetWhichHasShips.isPresent()){
					break;
				}
				universe.sendShips(1, closestOwnPlanetWhichHasShips.get().getId(), unknownPlanet.getId());
			}
		}
	}

}

File: src/main/kotlin/de/spries/fleetcommander/model/ai/behavior/SearchEnemyHomePlanetFleetStrategy.kt
package de.spries.fleetcommander.model.ai.behavior

import de.spries.fleetcommander.model.facade.PlayerSpecificUniverse

class SearchEnemyHomePlanetFleetStrategy : FleetStrategy {

    override fun sendShips(universe: PlayerSpecificUniverse) {
        val homePlanet = universe.getHomePlanet()
        val allPlanets = universe.getPlanets()
        val myPlanets = allPlanets.filter { p -> p.isInhabitedByMe() }
        val enemyHomePlanet = allPlanets.firstOrNull { p -> p.isKnownAsEnemyPlanet() && p.isHomePlanet() }
        val unknownPlanets = allPlanets
                .filter { p -> !p.isInhabitedByMe() && !p.isKnownAsEnemyPlanet() && p.getIncomingShipCount() == 0 }
                .sortedBy { it.distanceTo(homePlanet!!) }
                .toList()

        // Expand to next closest uninhabited planets
        for (unknownPlanet in unknownPlanets) {
            // send 1 ship from closest of my own planets which has ships > 0
            val closestOwnPlanetWhichHasShips = myPlanets
                    .sortedBy { it.distanceTo(unknownPlanet) }
                    .firstOrNull { p -> p.getShipCount() > 0 } ?: break

            universe.sendShips(1, closestOwnPlanetWhichHasShips.getId(), unknownPlanet.getId())
        }

        if (enemyHomePlanet != null) {
            // Full attack on enemy home planet
            myPlanets
                    .filter { p -> p.getShipCount() > 0 }
                    .forEach { p -> universe.sendShips(p.getShipCount(), p.getId(), enemyHomePlanet.getId()) }
        }
    }
}
